File: Problem_5_Longest_Palindromic_Substring/solution.py
def longest_palindromic_substring(s: str) -> str:
    if not s:
        return ""

    start = 0
    max_len = 1

    def expand(left: int, right: int) -> None:
        nonlocal start, max_len
        while left >= 0 and right < len(s) and s[left] == s[right]:
            curr_len = right - left + 1
            if curr_len > max_len:
                max_len = curr_len
                start = left
            left -= 1
            right += 1

    for i in range(len(s)):
        expand(i, i)
        expand(i, i + 1)

    return s[start:start + max_len]


if __name__ == "__main__":
    print(longest_palindromic_substring("babad"))
    print(longest_palindromic_substring("cbbd"))

File: Problem_5_Longest_Palindromic_Substring/solution.cpp
#include <string>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        if (s.size() < 2) {
            return s;
        }

        int start = 0;
        int maxLength = 1;

        for (int i = 0; i < static_cast<int>(s.size()); ++i) {
            int left = i;
            int right = i;
            while (left >= 0 && right < static_cast<int>(s.size()) && s[left] == s[right]) {
                int currentLength = right - left + 1;
                if (currentLength > maxLength) {
                    maxLength = currentLength;
                    start = left;
                }
                --left;
                ++right;
            }

            left = i;
            right = i + 1;
            while (left >= 0 && right < static_cast<int>(s.size()) && s[left] == s[right]) {
                int currentLength = right - left + 1;
                if (currentLength > maxLength) {
                    maxLength = currentLength;
                    start = left;
                }
                --left;
                ++right;
            }
        }

        return s.substr(start, maxLength);
    }
};
